### PR TITLE
Implemented methods for retrieving deals based on followed Users/Restaurants/Tags

### PR DESCRIPTION
**Pull request by lisasgoh**

### CODE REVIEW

File: src/main/java/com/google/step/datamanager/DealManagerDatastore.java
Comment (limli) [inline src/main/java/com/google/step/datamanager/DealManagerDatastore.java]: what's the point of `OLDEST_DEAL_TIMESTAMP` ? I think removing it won't affect the sorting

File: src/main/java/com/google/step/model/Deal.java
Comment (limli) [inline src/main/java/com/google/step/model/Deal.java]: this doesn't need to be an attribute, like how number of votes is not an attribute. It feels weird to suddenly have a private attribute here.

File: src/main/java/com/google/step/datamanager/DealManagerDatastore.java
Comment (limli) [inline src/main/java/com/google/step/datamanager/DealManagerDatastore.java]: why is this method `...RestaurantsOrUsers` ? Why only restaurant and user? Why is tags separate?
Maybe split all up into 3 different methods, or combine all

File: src/main/java/com/google/step/datamanager/DealManagerDatastore.java
Comment (limli) [inline src/main/java/com/google/step/datamanager/DealManagerDatastore.java]: don't call from the "Follow" table, use the FollowManager instead, or add methods into FollowManager if you need. Basically, don't have 2 classes managing the same table. Else, if say in the future we want to change the table format, then there will be too many things dependent on it.

File: src/main/java/com/google/step/datamanager/DealManagerDatastore.java
Comment (limli) [inline src/main/java/com/google/step/datamanager/DealManagerDatastore.java]: the complexity of this is increased by calling `getVotes` every time we want to compare, because `getVotes` is an expensive operation. Instead, call `getVotes` once for each deal

File: src/main/java/com/google/step/datamanager/DealManagerDatastore.java
Comment (limli) [inline src/main/java/com/google/step/datamanager/DealManagerDatastore.java]: use `LOCATION`?

File: src/test/java/com/google/step/datamanager/DealManagerDatastoreTest.java
Comment (limli) [inline src/test/java/com/google/step/datamanager/DealManagerDatastoreTest.java]: wait for #44 to be merged before merging this PR

**Review by limli — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/step/datamanager/DealManagerDatastore.java
Comment (lisasgoh) [reply src/main/java/com/google/step/datamanager/DealManagerDatastore.java]: The tags is kind of different since theres no need to search the datastore since there is the getDealIdsWithTag method  in dealTagManager but for restaurants/users, theres a need to search the datastore for the deal with the restaurantid/posterid value

File: src/main/java/com/google/step/datamanager/DealManagerDatastore.java
Comment (lisasgoh) [reply src/main/java/com/google/step/datamanager/DealManagerDatastore.java]: i just copied from reddit's hot algorithm lol which takes into account the datetime of the oldest post

File: src/main/java/com/google/step/datamanager/DealManagerDatastore.java
Comment (lisasgoh) [reply src/main/java/com/google/step/datamanager/DealManagerDatastore.java]: yea but the original method in follow manager is private so im not sure if i should make it public? since u said the people can tamper with the fieldnames

File: src/main/java/com/google/step/model/Deal.java
Comment (lisasgoh) [reply src/main/java/com/google/step/model/Deal.java:13]: but if its not an attribute, im not sure how to store the information? since it cant be a manager right

File: src/main/java/com/google/step/datamanager/DealManagerDatastore.java
Comment (limli) [reply src/main/java/com/google/step/datamanager/DealManagerDatastore.java]: what you're doing now also like tampering with field names
you can first call `getFollowedRestaurantIds(userId)` from `FollowManager` to get a `list` of restaurant IDs
Then lookup the Deal table to find restaurant IDs using `list`
That way you don't have to touch the Follow table at all.

File: src/main/java/com/google/step/datamanager/DealManagerDatastore.java
Comment (limli) [reply src/main/java/com/google/step/datamanager/DealManagerDatastore.java]: ok

File: src/main/java/com/google/step/datamanager/DealManagerDatastore.java
Comment (limli) [inline src/main/java/com/google/step/datamanager/DealManagerDatastore.java]: remove these

File: src/main/java/com/google/step/datamanager/DealManagerDatastore.java
Comment (limli) [inline src/main/java/com/google/step/datamanager/DealManagerDatastore.java]: you don't need to parse, you should be able to just compare the strings

File: src/main/java/com/google/step/datamanager/DealManagerDatastore.java
Comment (limli) [inline src/main/java/com/google/step/datamanager/DealManagerDatastore.java]: do you want to move this method to the servlet? Because the servlets will need to call getVotes anyway to print the deal details. It seems like there is a lot of extra calls to getVotes.

File: src/main/java/com/google/step/datamanager/DealManagerDatastore.java
Comment (limli) [inline src/main/java/com/google/step/datamanager/DealManagerDatastore.java]: How about instead of sorting the Map<String,Object> with casting everywhere, maybe can create a new private class of pair? Like
```
class DealPair<T extends Comparable<T>> implements Comparable<DealPair<T>> {

  public final T key;
  public final Deal deal;

  public DealPair(T key, Deal deal) {
    this.key = key;
    this.deal = deal;
  }

  @Override
  public int compareTo(DealPair<T> other) {
    return key.compareTo(other.key);
  }
}
```

Comment (lisasgoh) [reply src/main/java/com/google/step/datamanager/DealManagerDatastore.java]: ill do this for pr #67 cos i shifted over to the servlet

File: src/main/java/com/google/step/datamanager/DealManagerDatastore.java
Comment (lisasgoh) [reply src/main/java/com/google/step/datamanager/DealManagerDatastore.java]: will try this for pr #67 since i shifted this to servlet

File: src/main/java/com/google/step/datamanager/DealManagerDatastore.java
Comment (limli) [inline src/main/java/com/google/step/datamanager/DealManagerDatastore.java]: are these still used anywhere?

Comment (lisasgoh) [reply src/main/java/com/google/step/datamanager/DealManagerDatastore.java]: location is

**Review by jyx11011 — DISMISSED** [review]

**Review by limli — DISMISSED** [review]

File: src/main/java/com/google/step/datamanager/DealManagerDatastore.java
Comment (SuyashLakhotia) [inline src/main/java/com/google/step/datamanager/DealManagerDatastore.java]:
```suggestion
      Set<Long> ids, String filterAttribute) {
```

Don't need a long name here, the method name provides the context.

File: src/main/java/com/google/step/datamanager/DealManagerDatastore.java
Comment (SuyashLakhotia) [inline src/main/java/com/google/step/datamanager/DealManagerDatastore.java]: If you use `FilterOperator.IN` instead of `FilterOperator.EQUAL`, can you make this a single call to datastore?

File: src/main/webapp/js/user-page.js
Comment (SuyashLakhotia) [inline src/main/webapp/js/user-page.js]: Is this the right place for this change? Should this be a separate PR?

File: src/test/java/com/google/step/datamanager/DealManagerDatastoreTest.java
Comment (SuyashLakhotia) [inline src/test/java/com/google/step/datamanager/DealManagerDatastoreTest.java:248]:
```suggestion
    assertEquals(dealsForA.size(), 2);
```

to be consistent with your other assertions

File: src/test/java/com/google/step/datamanager/DealManagerDatastoreTest.java
Comment (SuyashLakhotia) [inline src/test/java/com/google/step/datamanager/DealManagerDatastoreTest.java]: Ditto.

File: src/test/java/com/google/step/datamanager/DealManagerDatastoreTest.java
Comment (SuyashLakhotia) [inline src/test/java/com/google/step/datamanager/DealManagerDatastoreTest.java:287]: Ditto.

File: src/test/java/com/google/step/datamanager/DealManagerDatastoreTest.java
Comment (SuyashLakhotia) [inline src/test/java/com/google/step/datamanager/DealManagerDatastoreTest.java:315]: Ditto.

**Review by SuyashLakhotia — CHANGES_REQUESTED** [review]

**Comment by limli** [comment]
change your base branch

File: src/test/java/com/google/step/datamanager/DealManagerDatastoreTest.java
Comment (lisasgoh) [reply src/test/java/com/google/step/datamanager/DealManagerDatastoreTest.java:248]: i fixed the other assertions instead since the first argument should be the expected value, and to be consistent with the other tests

**Review by SuyashLakhotia — APPROVED** [review]

**Pull request merged**
